### ISSUE TITLE
Logistic Regression - adit.io

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1055]
Logistic Regression - adit.io<br>
<br>
https://ift.tt/1V4XzJv<br>
<br>
<br>
<br>

<div>
<div id="content">
          
          <h1>Logistic Regression</h1>
<time>Written March 13, 2016</time>

<p>This post covers week 3 of Andrew Ng's excellent Coursera course on <a href="https://www.coursera.org/learn/machine-learning">machine learning</a>. You should read this post if the videos went too fast for you. To understand this post, you should know how linear regression works. If you don't, <a href="http://adit.io/posts/2016-02-20-Linear-Regression-in-Pictures.html">read this post</a> first.</p>

<p>Logistic "regression" is <strong>classification</strong> algorithm. You can use it to classify things:</p>

<p><img src="http://adit.io/imgs/logistic/oranges_vs_grapefruit.png"></p>

<p>So instead of trying to predict a number (what price should I sell my house for?), you are trying to classify something (hey is this a grapefruit?).</p>

<p>Just to recap, linear regression is simple: you take your data, and plot it on a graph. Then you draw a line that fits your data pretty well. Now you can use that line to predict things. So you could use this line to predict housing prices based on size:</p>

<p><img src="http://adit.io/imgs/regression/eyeballed_line_good.png"></p>

<p>The best line is the one that minimizes a cost. For linear regression, the cost was the distance from each point of data to the line.</p>

<p>Similarly, when you are trying to classify something, first you plot your data on a graph.</p>

<p>Suppose you are trying to classify a new piece of fruit as an orange or a grapefruit. Here's your existing data:</p>

<p><img src="http://adit.io/imgs/logistic/grapefruit_graphed.png"></p>

<p>i.e. anything that is less than 4 inches in diameter is an orange, anything greater than that is a grapefruit.</p>

<p>Now you can draw a line in this image. Anything to the left of the line is an orange, and anything to the right is a grapefruit:</p>

<p><img src="http://adit.io/imgs/logistic/grapefruit_graphed_with_line.png"></p>

<p>This is how logistic regression works. You plot your data, then draw a line. This line is called a <strong>decision boundary</strong>, and it splits the data into different classes (in this case, oranges and grapefruit).</p>

<p>Just like linear regression, logistic regression involves trying to find a line: except in this case, it is the decision boundary.</p>
<h2 id="prediction-function">Prediction function</h2>
<p>For the linear regression example, our data looked like this:</p>

<p><img src="http://adit.io/imgs/logistic/linear_regression_data.png"></p>

<p>i.e. <strong>home size -&gt; price of home</strong>.</p>

<p>For this example, our data looks like this:</p>

<p><img src="http://adit.io/imgs/logistic/logistic_regression_data.png"></p>

<p>i.e. <strong>fruit diameter -&gt; is it a grapefruit</strong>. If the fruit is a grapefruit, <code>y = 1</code>. Otherwise, <code>y = 0</code>.</p>

<p>In the linear regression example, we predicted a number (like a 3000 sq ft home will sell for $270,000). For logistic regression, we predict a probability, like "there's a 90% chance that this is a grapefruit".</p>

<p>In linear regression, the prediction formula looked like this:</p>

<p><img src="http://adit.io/imgs/regression/h_x_formula.png"></p>

<p>You would calculate values for <code>theta_0</code> and <code>theta_1</code>, and then use them to make a prediction. For example, if we wanted to know how much to sell a 3000 sq foot house for:</p>

<p><img src="http://adit.io/imgs/logistic/linear_regression_3000.png"></p>

<p>Once you had the optimal values for <code>theta_0</code> and <code>theta_1</code>, the prediction part was easy.</p>

<p>We use almost the same formula for logistic regression, with one change:</p>

<p><img src="http://adit.io/imgs/logistic/prediction_formula.png"></p>

<p>Remember, <strong>the prediction function for logistic regression spits out a percentage</strong>. As in "we are 95% sure this is a grapefruit". The <code>g(...)</code> is called the <strong>sigmoid</strong> function. This function is what makes the prediction function output a percentage. <strong>The sigmoid function constrains the result to between 0 and 1</strong>. So a result of .95 would mean we are 95% certain of something. That's all you really need to know about the sigmoid function.</p>

<p>The sigmoid function is:</p>

<p><img src="http://adit.io/imgs/logistic/sigmoid.png"></p>

<p>Just like linear regression, we will use a cost function to find good values for <code>theta_0</code> and <code>theta_1</code>. But for our first example, I already know two good values. So lets plug them in:</p>

<p><img src="http://adit.io/imgs/logistic/formula_with_thetas.png"></p>

<p><code>theta_0 = -7</code> and <code>theta_1 = 2</code>. Now we can use this function to predict whether a fruit is a grapefruit! Here are the results, for fruits of sizes 1 to 5 inches in diameter:</p>

<p><img src="http://adit.io/imgs/logistic/data_with_percentages.png"></p>

<p>You can try this yourself by just <a href="https://www.wolframalpha.com/input/?i=sigmoid(-7+%2B+2*2)">plugging the numbers into WolframAlpha</a>.</p>

<p>Remember, that percentage is your confidence that this fruit is a grapefruit. That's pretty close! The line looks like this:</p>

<p><img src="http://adit.io/imgs/logistic/line_graphed.png"></p>

<p>That's on the nose. Anything to the left of the line is an orange, and anything to the right is a grapefruit. Remember, the sigmoid function converts that into probabilities. The farther left you get, the more lower the percentage that the fruit is a grapefruit. So the sigmoid function smooths out the answers, instead of just spitting out a yes or a no:</p>

<p><img src="http://adit.io/imgs/logistic/line_graphed_sigmoid.png"></p>

<p>So now that we have seen the prediction function in action, lets see how to find those optimal values for <code>theta_0</code> and <code>theta_1</code>. Just to recap, the only thing we have changed so far is the prediction function <code>h(x)</code>, to add the sigmoid function. Everything else is still the same as linear regression. The next thing we need to change is the cost function.</p>
<h2 id="the-cost-function">The cost function</h2>
<p>In linear regression, we calculated the cost of using a line:</p>

<p><img src="http://adit.io/imgs/regression/good_vs_bad_fit.png"></p>

<p>The cost was based on how far off the line was from our data points.</p>

<p>In logistic regression, the cost again depends on how far off our predictions are from our actual data. But we are using percentages, so the cost is calculated a little differently.</p>

<p>For example, if we say "There's a 95% chance that this is a grapefruit", but it turns out to be an orange, we should get penalized heavily (i.e. the cost should be higher). But if we say "There's a 55% chance that this is a grapefruit" and it turns to be an orange, that should be a lower penalty (i.e. the cost should be lower).</p>

<p>So the cost for logistic regression is calculated based on how far off your probability was. We calculate it using a log scale, like this:</p>

<p><img src="http://adit.io/imgs/logistic/log_graph.png"></p>

<p><code>y</code> is the actual result. So if <code>y = 1</code> i.e. it was a grapefruit, and you predicted 1 (i.e. 100% probability it is a grapefruit), there's no penalty. But if you predicted 0 (0% probability it is a grapefruit), then you get penalized heavily.</p>

<p>The wronger you are, the more you get penalized. We use the <code>log</code> function for this.</p>

<p>The cost function for <strong>linear regression</strong> was:</p>

<p><img src="http://adit.io/imgs/regression/cost_function_math_formula.png"></p>

<p>The cost function for <strong>logistic regression</strong> is:</p>

<p><img src="http://adit.io/imgs/logistic/cost_function_logistic_regression.png"></p>

<p>And there's a clever way to write that on one line, like this:</p>

<p><img src="http://adit.io/imgs/logistic/cost_function_one_liner.png"></p>

<p>This works because one of those two will always be zero, so only the other one will get used, just like in the <code>if</code> statement. I personally prefer the <code>if</code> statement approach, but the videos use the one-liner so I've included it here.</p>

<p>Here it is in Octave:</p>
<div><pre><code><span>function</span> <span>distance</span> <span>=</span> <span>cost</span><span>(</span><span>theta_0</span><span>,</span> <span>theta_1</span><span>,</span> <span>x</span><span>,</span> <span>y</span><span>)</span>
  <span>distance</span> <span>=</span> <span>0</span><span>;</span>
  <span>for</span> <span>i</span> <span>=</span> <span>1</span><span>:</span><span>length</span><span>(</span><span>x</span><span>)</span> <span>%</span> <span>arrays</span> <span>in</span> <span>octave</span> <span>are</span> <span>indexed</span> <span>starting</span> <span>at</span> <span>1</span>
    <span>%</span> <span>you</span> <span>calculate</span> <span>the</span> <span>cost</span> <span>differently</span> <span>based</span> <span>on</span> <span>whether</span> <span>the</span> <span>actual</span> <span>value</span> <span>is</span> <span>1</span> <span>or</span> <span>0</span><span>.</span>
    <span>if</span> <span>(</span><span>y</span><span>(</span><span>i</span><span>)</span> <span>==</span> <span>1</span><span>)</span>
      <span>distance</span> <span>+=</span> <span>-</span><span>log</span><span>(</span><span>h</span><span>(</span><span>x</span><span>(</span><span>i</span><span>),</span> <span>theta_0</span><span>,</span> <span>theta_1</span><span>));</span>
    <span>else</span>
      <span>distance</span> <span>+=</span> <span>-</span><span>log</span><span>(</span><span>1</span> <span>-</span> <span>h</span><span>(</span><span>x</span><span>(</span><span>i</span><span>),</span> <span>theta_0</span><span>,</span> <span>theta_1</span><span>));</span>
    <span>end</span>
  <span>end</span>
  <span>%</span> <span>get</span> <span>how</span> <span>far</span> <span>off</span> <span>we</span> <span>were</span> <span>on</span> <span>average</span>
  <span>distance</span> <span>=</span> <span>distance</span> <span>/</span> <span>length</span><span>(</span><span>x</span><span>);</span>
<span>end</span>
</code></pre></div>
<p>To recap:</p>

<ul>
<li>We added the sigmoid function to the prediction function, so it outputs percentages</li>
<li>We changed the cost function (used the log function instead of predicted - actual)</li>
</ul>

<p>That's it!  Those two things are different, everything else is the same. We still use gradient descent to iteratively find the correct values for theta (and even the gradient descent formula still stays the same).</p>

<p><a href="https://gist.github.com/egonSchiele/54abdc69ad843731eb77">Here's the full example in Octave</a>.</p>

<p>Now you know how to build a logistic regression classifier!</p>

<p><img src="http://adit.io/imgs/logistic/grocery_store.png"></p>
<h2 id="multi-class-classification">Multi-class classification</h2>
<p>You know how logistic regression works now. The rest of this post is optional, and talks about some things you will run into when using logistic regression in practice.</p>

<p>How would you use logistic regression to classify between many different types of fruit, not just two different types?</p>

<p><img src="http://adit.io/imgs/logistic/different_fruit.png"></p>

<p>(O = orange, G = grapefruit, A = apple).</p>

<p>Then you need multiple decision boundaries:</p>

<p><img src="http://adit.io/imgs/logistic/multiple_decision_boundaries.png"></p>

<p>Turns out there's a really easy way to do this using something called one-vs-all classification! Here's the idea. First, just concentrate on grapefruit:</p>

<p><img src="http://adit.io/imgs/logistic/just_grapefruit.png"></p>

<p>So either the fruit IS an grapefruit, or ISN'T an grapefruit (oranges and apples temporarily get lumped together). You do this classification and get a numbe, lets say 25%. So there's a 25% chance this is a grapefruit, and <em>there's a 75% chance this is something other than a grapefruit</em>.</p>

<p>Next you do the same with oranges:</p>

<p><img src="http://adit.io/imgs/logistic/just_oranges.png"></p>

<p>So either it IS an orange, or ISN'T (grapefruits and apples get lumped together). The output is 75%.</p>

<p>Finally, do this for apples:</p>

<p><img src="http://adit.io/imgs/logistic/just_apples.png"></p>

<p>And you get: 30%.</p>

<p>So there's a 25% chance this is a grapefruit, 75% chance this is an orange, and 30% chance this is an apple. The result: it is an orange!</p>
<h2 id="non-linear-classification">Non-linear classification</h2>
<p>Suppose your data set looks like this:</p>

<p><img src="http://adit.io/imgs/logistic/non_linear_data.png"></p>

<p>There's no straight line that you can use as your decision boundary. You need a non-linear boundary, like this:</p>

<p><img src="http://adit.io/imgs/logistic/non_linear_boundary.png"></p>

<p>You can do this with logistic regression <a href="https://gist.github.com/egonSchiele/979922cfe7966033f209">like this</a>.</p>

<p>It uses a slightly different formula for prediction:</p>

<p><img src="http://adit.io/imgs/logistic/non_linear_prediction_formula.png"></p>

<p>But in general it is really hard to do non-linear classification using logistic regression. Logistic regression is not good for something like this:</p>

<p><img src="http://adit.io/imgs/logistic/extreme_non_linear.png"></p>

<p>This is where non-linear classifiers shine. Neural networks and support vector machines (SVMs) are both good at non-linear classification.</p>

<p><img src="http://adit.io/imgs/logistic/classification_machine.png"></p>
<h2 id="recap">Recap</h2>
<ul>
<li>The logistic function spits out a percentage</li>
<li>The sigmoid function is used to constrain the output to between 0 and 1</li>
<li>The cost is calculated using a log scale: the more wrong you were, the more you get penalized</li>
<li>A decision boundary is a line you draw to separate your data into two different classes</li>
<li>If you have multiple classes, use one-vs-all classification</li>
<li>If you need non-linear classification, choose neural networks or support vector machines</li>
</ul>

<p id="disqus_thread"></p>

Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a>

<a href="http://disqus.com">blog comments powered by <span>Disqus</span></a>

        </div>
      </div><br>
<br>
<br>
<br>
<br>
<br>
via adit.io http://adit.io<br>
<br>
October 9, 2018 at 04:47PM
